### PR TITLE
[JAX] Use TE with_sharding_constraint wrapper even if flax returns successful logical sharding rules

### LINKED ISSUE #2500: [JAX] Sharding constraint for rng_state in fused_attn_fwd makes it is impossible to use in shard_map
**Is your feature request related to a problem? Please describe.**

Hey guys, We have a custom transformer implementation with some parts wrapped in shard_map but using with_sharding_constraint here
https://github.com/NVIDIA/TransformerEngine/blob/5afbb0e14f58e068a6370eb9e5dbcbb96bc7ac04/transformer_engine/jax/cpp_extensions/attention.py#L3446

makes it impossible to use fused_attn_fwd in shard_map. And this is basically the only row that makes it impossible. To solve this issue, I just copy-paste source code for the following functions:
```
fused_attn_fwd
_fused_attn_fwd_rule
_fused_attn
fused_attn
```
while remove this particular row and it makes it possible to properly use cudnn attention in shard_map. 

**Describe the solution you'd like**

I wonder if we can move this sharding constraint somewhere else, outside of fused attention fwd?

### PR DESCRIPTION
**Pull request by KshitijLakhani**
# Description

Fix JAX logical-axis sharding constraints under `shard_map`. When Flax logical axis rules were present, `with_sharding_constraint_by_logical_axes` called `flax.linen.with_logical_constraint` directly. That bypassed TE’s `with_sharding_constraint` wrapper, which strips mesh axes already owned by `shard_map`. As a result, fused attention paths such as `softmax_offset` could apply constraints on manual axes and fail inside `shard_map`.
This change resolves Flax logical axes to a `PartitionSpec` with `flax.linen.logical_to_mesh_axes`, then routes the constraint through TE’s shard-map-aware wrapper.

Fixes #2500

# Reproducer

Minimal reproducer for the softmax_offset path. This mirrors fused attention’s (1, heads, 1, 1) tensor and applies the same logical-axis constraint under shard_map. Expected after fix: (1, 8, 1, 1)

```
  from functools import partial
  import flax.linen as nn
  import jax
  import jax.numpy as jnp
  import numpy as np
  from jax.sharding import Mesh
  from jax.sharding import PartitionSpec as P
  from transformer_engine.jax.sharding import (
      HEAD_AXES,
      with_sharding_constraint_by_logical_axes,
  )
  mesh = Mesh(np.array(jax.devices()[:2]), ("x",))
  @partial(
      jax.shard_map,
      mesh=mesh,
      in_specs=P(None, "x", None, None),
      out_specs=P(None, "x", None, None),
  )
  def constrain_softmax_offset(softmax_offset):
      return with_sharding_constraint_by_logical_axes(
          softmax_offset, (None, HEAD_AXES, None, None)
      )
  with nn.logical_axis_rules(((HEAD_AXES, "x"),)):
      out = constrain_softmax_offset(jnp.ones((1, 8, 1, 1), dtype=jnp.float32))
  print(out.shape)
```
  
## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [X] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by jberchtold-nvidia — APPROVED** [review]
Good catch, LGTM, thanks!
